File: src/lib/api/narrativeText.ts
import { call } from './call'
import type {
  Condition,
  DialogueSlot,
  GenerationPolicy,
  NarrativeDiagnostic,
  Participant,
  Precondition,
  SceneId,
  Stamp,
  TextAssetId,
  TextEntryId,
} from './narrative'

export type { TextAssetId, TextEntryId } from './narrative'

/**
 * The Text library's command surface: barks, ambient exchanges, companion
 * reactions, codex entries, quest summaries and journals (#167).
 *
 * The same two spellings meet here as in `narrative.ts` and for the same
 * reason: `TextCatalog` and `TextFile` are envelopes and are camelCase, while
 * `TextAsset` and everything under it is the YAML in `narrative/texts/*.yaml`
 * verbatim and is snake_case. A camelCase mirror would give the form editor and
 * the Source view two vocabularies for one file.
 *
 * The load-bearing reuse is `DialogueSlot`: a bark's line is the *same* type as
 * a scene line, so the review, freshness and policy controls that already know
 * how to render one need no supporting-text special case. If this file ever
 * grows its own line type, that stops being true for a quarter of the words in
 * the game.
 */

/* ── domain ───────────────────────────────────────────────────────────────── */

/** Which of the six supporting kinds an asset is. */
export type TextKind = 'bark' | 'ambient' | 'reaction' | 'codex' | 'quest_summary' | 'journal'

/**
 * How the runtime picks between eligible entries, and what the next trigger
 * does. Authored rather than left to the host: two hosts choosing differently
 * would make the same project play differently.
 */
export type RepeatPolicy =
  /** Always the first eligible entry — prose whose wording is a function of state. */
  | 'first'
  /** Each eligible entry once, in author order, then nothing. */
  | 'once'
  /** Round-robin in author order. */
  | 'cycle'
  /** A seeded bag: each eligible entry once per round, never repeating across the join. */
  | 'shuffle'

/** An authored record this text is about, and a dependency edge for #168. */
export type SourceLink =
  | { scene: SceneId }
  | { quest: string }
  | { fact: string }
  | { event: string }
  | { character: string }

/** The host-side moment an asset belongs to. */
export interface Trigger {
  /** A declared identifier, like a host command name: `[a-z][a-z0-9_]*`. */
  event: string
  /** Absent is unconditional, which is not the same as `never`. */
  when?: Condition
}

/** One alternative delivery: a single bark, a version of an exchange, a page. */
export interface TextEntry {
  id: TextEntryId
  label?: string
  when?: Condition
  /** Ordered. More than one only for an exchange or a passage of prose. */
  lines?: DialogueSlot[]
}

export interface TextAsset {
  id: TextAssetId
  kind: TextKind
  name: string
  summary?: string
  trigger: Trigger
  sources?: SourceLink[]
  /** Empty for standalone prose; every entity that speaks has to be listed. */
  participants?: Participant[]
  must_convey?: string[]
  must_not_reveal?: string[]
  repeat?: RepeatPolicy
  /** Whether generation may write into this asset at all. */
  policy?: GenerationPolicy
  entries?: TextEntry[]
  tombstones?: unknown[]
}

/* ── envelopes ────────────────────────────────────────────────────────────── */

export interface TextSummary {
  id: TextAssetId
  kind: TextKind
  name: string
  /** The file name stem, minted at create and stable across a rename. */
  slug: string
  /** Project-relative, `/`-separated. */
  rel: string
}

export interface TextCatalog {
  assets: TextSummary[]
  /** Files in the directory that could not be identified, listed not dropped. */
  unreadable: { rel: string; reason: string }[]
}

/** One asset, whole, with the precondition for writing it back. */
export interface TextFile {
  asset: TextAsset
  slug: string
  rel: string
  /** `null` means "we believe there is no file". */
  stamp: Stamp | null
}

/* ── commands ─────────────────────────────────────────────────────────────── */

export const narrativeTexts = () => call<TextCatalog>('narrative_texts')

export const narrativeTextGet = (assetId: TextAssetId) =>
  call<TextFile>('narrative_text_get', { assetId })

/**
 * Create an asset from a kind template.
 *
 * `event` is required rather than defaulted: an asset with no trigger is one
 * the host can never ask for, and a placeholder name would be a
 * working-looking binding to a moment that does not exist.
 */
export const narrativeTextCreate = (kind: TextKind, name: string, event: string) =>
  call<TextFile>('narrative_text_create', { kind, name, event })

/**
 * Write a whole supporting text document.
 *
 * One write path, guarded by the same precondition a scene uses, so a stale
 * editor tab cannot overwrite a colleague's newer file. `slug` is only
 * consulted for an asset the project does not currently have.
 */
export const narrativeTextSave = (asset: TextAsset, expected: Precondition, slug?: string | null) =>
  call<TextFile>('narrative_text_save', { asset, expected, slug: slug ?? null })

export const narrativeTextDelete = (assetId: TextAssetId) =>
  call<void>('narrative_text_delete', { assetId })

/**
 * What is wrong with one asset. Pass `asset` to diagnose the document as the
 * writer currently has it, unsaved edits and all.
 */
export const narrativeTextDiagnostics = (assetId: TextAssetId, asset?: TextAsset | null) =>
  call<NarrativeDiagnostic[]>('narrative_text_diagnostics', { assetId, asset: asset ?? null })

File: src-tauri/src/commands/narrative_text.rs
//! The Text library's commands: supporting text assets on the same terms as
//! scenes (#167).
//!
//! Everything here is the scene surface in `commands::narrative` with the
//! scene-shaped parts removed, and that is deliberate rather than lazy. The
//! guard is the same [`Precondition`], the write is the same whole-document
//! guarded save, and the diagnostics are the same [`DiagnosticView`], because a
//! bark is authored, reviewed, merged and recovered by the same people under
//! the same rules. A second, gentler write path for supporting text would be a
//! second place a collaborator's work could be silently overwritten, and it
//! would be discovered by somebody losing a morning's barks.
//!
//! What is missing compared with scenes, and why:
//!
//! - **No layout.** A text asset has no canvas, which is the same absence that
//!   keeps `Destination` out of its model.
//! - **No rename-in-place command.** A scene has one because its title is shown
//!   in a dozen places; an asset's name is edited in the same form as the rest
//!   of it and arrives through the ordinary save.
//! - **No library index.** The catalog is read from the folder on every call.
//!   Indexing it is the same work as adding it to the scene library's search,
//!   and belongs with that rather than half-done here.

use serde::Serialize;
use tauri::State;
use wobu_narrative::{Name, TextAsset, TextAssetId, TextKind};
use wobu_store::atomic::Stamp;
use wobu_store::{Project, SourceSave, TextFile};

use super::narrative::{DiagnosticView, Precondition, UnreadableScene};
use crate::error::{CommandResult, WobuError};
use crate::state::AppState;

/// One supporting text asset as the library lists it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TextSummary {
    pub id: TextAssetId,
    pub kind: TextKind,
    pub name: String,
    pub slug: String,
    pub rel: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TextCatalogView {
    pub assets: Vec<TextSummary>,
    pub unreadable: Vec<UnreadableScene>,
}

/// One asset, its file and the stamp a save has to present back.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TextFileView {
    pub asset: TextAsset,
    pub slug: String,
    pub rel: String,
    pub stamp: Option<Stamp>,
}

impl TextFileView {
    fn of(file: &TextFile) -> TextFileView {
        TextFileView {
            asset: file.asset.clone(),
            slug: file.slug().to_string(),
            rel: file.rel.clone(),
            stamp: file.stamp.clone(),
        }
    }
}

#[tauri::command]
pub fn narrative_texts(state: State<'_, AppState>) -> CommandResult<TextCatalogView> {
    state.with(|project| {
        let catalog = project.text_catalog()?;
        Ok(TextCatalogView {
            assets: catalog
                .assets
                .iter()
                .map(|entry| TextSummary {
                    id: entry.id,
                    kind: entry.kind,
                    name: entry.name.clone(),
                    slug: entry.slug.clone(),
                    rel: entry.rel.clone(),
                })
                .collect(),
            unreadable: catalog
                .unreadable
                .iter()
                .map(|entry| UnreadableScene {
                    rel: entry.rel.clone(),
                    reason: entry.reason.clone(),
                })
                .collect(),
        })
    })
}

#[tauri::command]
pub fn narrative_text_get(
    state: State<'_, AppState>,
    asset_id: TextAssetId,
) -> CommandResult<TextFileView> {
    state.with(|project| Ok(TextFileView::of(&project.load_text_asset(asset_id)?)))
}

/// Create an asset from a kind template.
///
/// `event` is required rather than defaulted, because an asset with no trigger
/// is an asset the host can never ask for, and a placeholder event name would
/// be a working-looking binding to a moment that does not exist. The kind
/// decides the starting repeat policy — see `TextKind::default_repeat`.
#[tauri::command]
pub fn narrative_text_create(
    state: State<'_, AppState>,
    kind: TextKind,
    name: String,
    event: String,
) -> CommandResult<TextFileView> {
    state.with(|project| {
        let event = Name::new(event).map_err(|error| {
            WobuError::new(crate::error::Code::Invalid, "That is not a valid trigger name.")
                .with_detail(error.to_string())
        })?;
        Ok(TextFileView::of(&project.create_text_asset(kind, &name, event)?))
    })
}

#[tauri::command]
pub fn narrative_text_save(
    state: State<'_, AppState>,
    asset: TextAsset,
    slug: Option<String>,
    expected: Precondition,
) -> CommandResult<TextFileView> {
    state.with(|project| save_text(project, asset, slug.as_deref(), &expected))
}

fn save_text(
    project: &mut Project,
    asset: TextAsset,
    slug: Option<&str>,
    expected: &Precondition,
) -> CommandResult<TextFileView> {
    if matches!(expected, Precondition::Current) {
        return Err(WobuError::new(
            crate::error::Code::Invalid,
            "Text saves require the original stamp, so a stale editor cannot overwrite a newer file.",
        ));
    }
    let catalog = project.text_catalog()?;
    let rel = match catalog.find(asset.id) {
        // The catalog is authoritative for an asset that exists: taking the path
        // from the caller would let a stale editor tab write to wherever it last
        // remembered the file being.
        Some(entry) => entry.rel.clone(),
        None => {
            let taken = catalog.slugs();
            let base = wobu_core::slugify(slug.unwrap_or(&asset.name))?;
            let unique = wobu_core::unique_slug(&base, &|candidate| taken.contains(candidate));
            wobu_store::narrative::text_rel(&unique)
        }
    };

    // Read without parsing: the precondition is a fact about the bytes, and a
    // file somebody broke in a text editor still has a stamp — otherwise the
    // person fixing it could not save over it.
    let path = wobu_store::paths::from_rel_string(project.root(), &rel);
    let on_disk = wobu_store::atomic::read_stamped(&path)?.map(|(_, stamp)| stamp);

    let mut file = TextFile { asset, rel, stamp: expected.against(on_disk) };
    match project.save_text_asset(&mut file)? {
        SourceSave::Saved(_) => Ok(TextFileView::of(&file)),
        SourceSave::Conflict { conflict_path } => Err(WobuError::conflict(conflict_path)),
    }
}

#[tauri::command]
pub fn narrative_text_delete(
    state: State<'_, AppState>,
    asset_id: TextAssetId,
) -> CommandResult<()> {
    state.with(|project| Ok(project.delete_text_asset(asset_id)?))
}

/// What is wrong with one asset, keyed by the element responsible.
///
/// `asset` overrides what is on disk for the same reason the scene command
/// accepts one: the editor holds unsaved edits, and a diagnostics call that
/// could only read the saved file would report problems the writer has already
/// fixed. The state schema always comes from the project, so a condition is
/// checked against the variables that really exist.
#[tauri::command]
pub fn narrative_text_diagnostics(
    state: State<'_, AppState>,
    asset_id: TextAssetId,
    asset: Option<TextAsset>,
) -> CommandResult<Vec<DiagnosticView>> {
    state.with(|project| {
        let asset = match asset {
            Some(asset) => asset,
            None => project.load_text_asset(asset_id)?.asset,
        };
        let schema = project.state_schema()?;
        Ok(asset.diagnostics(&schema).iter().map(DiagnosticView::of).collect())
    })
}

#[cfg(test)]
mod tests;
